### PR TITLE
Algo: Add CCTK_HOST CCTK_DEVICE to function brent and newton_raphson_nd

### LINKED ISSUE #108: Add CCTK_HOST and CCTK_DEVICE for functions in Algo
Should we add CCTK_HOST and CCTK_DEVICE in front of those functions in Algo?
In AsterX/Con2PrimFactory, we called function `brent()` which requires us to add CCTK_HOST and CCTK_DEVICE to make it work with GPU.

### PR DESCRIPTION
**Pull request by lwJi**
Closes #108.

### CODE REVIEW

**Comment by eschnett** [comment]
@lwJi Did you try using these functions in a GPU kernel? Can you set up a test case that would call them? There does not need to be a run-time test (because we currently cannot run tests on GPUs), but compile time tests would be good.

**Comment by lwJi** [comment]
I have tried `brent()` in a GPU kernel and it works perfectly. I haven't test `newton_raphson` yet. And the reason I haven't add CCTK_DEVICE for the other functions is they are still using `boost::`.

Aha. I misunderstood what you said before. I will try to add a `Test_roots`.

**Comment by lwJi** [comment]
@eschnett I think the tests are already exist in `root.cxx`
https://github.com/eschnett/CarpetX/blob/b68c3ae7fc64323bf0a2b007a3db4624b5d30f50/Algo/src/roots.cxx#L66

**Comment by eschnett** [comment]
These functions are built for the CPU only, not for a GPU.

**Comment by lwJi** [comment]
Now I get it (finally). Let me try again ;)

**Comment by lwJi** [comment]
@eschnett I think I have finished adding GPU tests (for `brent` and `newton_raphson_nd` only for the moment). Please let me know if I missed anything else

File: Algo/test/test_roots.par [hunk 1]
```diff
@@ -1,2 +1,8 @@
-ActiveThorns = "Algo"
+ActiveThorns = "
+    CarpetX
+    IOUtil
+    Algo
```
Comment (eschnett) [inline Algo/test/test_roots.par:4]: Can you sort the thorn names alphabetically?

File: Algo/src/roots.cxx [hunk 2]
```diff
@@ -28,9 +29,11 @@ template <typename T> std::pair<vec<T, 2>, mat<T, 2> > gnd(vec<T, 2> x) {
 extern "C" void Test_roots(CCTK_ARGUMENTS) {
   DECLARE_CCTK_ARGUMENTS;
 
-  auto fn = [](auto x) { return x * x - 2; };
-  auto fnd = [](auto x) { return std::make_tuple(x * x - 2, 2 * x); };
-  auto fnd2 = [](auto x) { return std::make_tuple(x * x - 2, 2 * x, 2); };
+  auto fn = []CCTK_HOST CCTK_DEVICE (double x) { return x * x - 2; };
```
Comment (eschnett) [inline Algo/src/roots.cxx:32]: Can you use `auto` (or `CCTK_REAL`) instead of `double`?

File: Algo/src/roots.cxx [hunk 3]
```diff
@@ -115,21 +105,45 @@ extern "C" void Test_roots(CCTK_ARGUMENTS) {
...
-                               vec<CCTK_REAL, 2>{10.0, 10.0}, minbits, maxiters,
-                               iters, failed);
-    // CCTK_VINFO("maxiters=%d iters=%d", maxiters, iters);
-    // CCTK_VINFO("lo=%.17g hi=%.17g", double(lo), double(hi));
-    assert(iters < maxiters);
-    assert(sumabs(gn(x)) < 1.0e-9);
-    CCTK_VINFO("Test_newton_raphson_nd succeeded in %d iterations", iters);
+    const amrex::Box box(amrex::IntVect(0), amrex::IntVect(0, 0, 0));
+
+    amrex::ParallelFor(
```
Comment (eschnett) [inline Algo/src/roots.cxx:110]: Use `amrex::launch` instead; see https://amrex-codes.github.io/amrex/docs_html/GPU.html#launching-general-kernels .

File: Algo/src/roots.cxx [hunk 4]
```diff
@@ -115,21 +105,45 @@ extern "C" void Test_roots(CCTK_ARGUMENTS) {
...
+      const int maxiters = 100;
+      int iters;
+      auto [lo, hi] = brent(fn, 1.0, 2.0, minbits, maxiters, iters);
+      // CCTK_VINFO("maxiters=%d iters=%d", maxiters, iters);
+      // CCTK_VINFO("lo=%.17g hi=%.17g", double(lo), double(hi));
+      assert(iters < maxiters);
+      assert(hi >= lo && hi - lo <= std::scalbn(2.0, -minbits));
+      assert(fn(lo) <= 0 && fn(hi) >= 0);
+      //CCTK_VINFO("Test_brent succeeded in %d iterations", iters);
+      printf("INFO (Algo): Test_brent succeeded in %d iterations\n", iters);
```
Comment (eschnett) [inline Algo/src/roots.cxx:123]: Don't call `printf` from inside the kernel. Instead, check whether the kernel finished successfully outside the kernel, and then call `CCTK_INFO`.

File: Algo/src/roots.cxx [hunk 5]
```diff
@@ -16,10 +16,11 @@ namespace Algo {
 using Arith::mat, Arith::vec;
 
 namespace {
-template <typename T> vec<T, 2> gn(vec<T, 2> x) {
+template <typename T> CCTK_HOST CCTK_DEVICE vec<T, 2> gn(vec<T, 2> x) {
```
Comment (eschnett) [inline Algo/src/roots.cxx:19]: If these functions are only called inside a kernel then the `CCTK_HOST` declaration is not necessary.

File: Algo/src/roots.hxx [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #ifndef ALGO_HXX
 #define ALGO_HXX
 
+#include "loop.hxx"
```
Comment (eschnett) [inline Algo/src/roots.hxx:4]: Can you write `<loop.hxx>`, and sort the statements alphabetically?

**Review by eschnett — CHANGES_REQUESTED** [review]

File: Algo/src/roots.cxx (hunk 2 above)
Comment (lwJi) [reply Algo/src/roots.cxx:32]: should I change all the `double` to `CCTK_REAL` which appears in the file

**Comment by lwJi** [comment]
@eschnett I have change the PR accordingly, please take another look. Thanks

File: Algo/src/roots.cxx [hunk 7]
```diff
@@ -114,22 +104,50 @@ extern "C" void Test_roots(CCTK_ARGUMENTS) {
...
-    auto x = newton_raphson_nd(gnd<CCTK_REAL>, vec<CCTK_REAL, 2>{1.0, 1.0},
-                               vec<CCTK_REAL, 2>{0.0, 0.0},
-                               vec<CCTK_REAL, 2>{10.0, 10.0}, minbits, maxiters,
-                               iters, failed);
-    // CCTK_VINFO("maxiters=%d iters=%d", maxiters, iters);
-    // CCTK_VINFO("lo=%.17g hi=%.17g", double(lo), double(hi));
-    assert(iters < maxiters);
-    assert(sumabs(gn(x)) < 1.0e-9);
-    CCTK_VINFO("Test_newton_raphson_nd succeeded in %d iterations", iters);
+    amrex::launch(box, [=] CCTK_DEVICE(amrex::Box const& tbx)
```
Comment (eschnett) [inline Algo/src/roots.cxx:112]: I think that launching a GPU kernel does not wait for the kernel to finish running. Can you check, and add the respective wait statement if necessary?

**Review by eschnett — CHANGES_REQUESTED** [review]

File: Algo/src/roots.cxx (hunk 7 above)
Comment (lwJi) [reply Algo/src/roots.cxx:112]: there is an implicit sync in `copyToHost`
```
    T* copyToHost ()
    {
#ifdef AMREX_USE_GPU
        if (d_data)
        {
            Gpu::dtoh_memcpy_async(h_data, d_data, m_size*sizeof(T));
            Gpu::streamSynchronize();
        }
#endif
        return h_data;
    }
```

**Review by eschnett — APPROVED** [review]
